File: lc_56_merge-intervals.js
/**
 * @param {number[][]} intervals
 * @return {number[][]}
 */
var merge = function (intervals) {
  const output = [];
  intervals.sort((a, b) => a[0] - b[0]);
  let previousStartTime = intervals[0][0];
  let previousEndTime = intervals[0][1];

  for (let i = 1; i < intervals.length; i++) {
    const currentInterval = intervals[i];
    if (currentInterval[0] - previousEndTime > 0) {
      output.push([previousStartTime, previousEndTime]);
      previousStartTime = currentInterval[0];
      previousEndTime = currentInterval[1];
    } else {
      previousEndTime = Math.max(previousEndTime, currentInterval[1]);
    }
  }
  output.push([previousStartTime, previousEndTime]);
  return output;
};

File: lc_56_merge-intervals.java
class Solution {
  public int[][] merge(int[][] intervals) {
    Arrays.sort(intervals, (a, b) -> Integer.compare(a[0], b[0]));
    int previousStartTime = intervals[0][0];
    int previousEndTime = intervals[0][1];
    ArrayList<int[]> output = new ArrayList<>();
    for (int i = 1; i < intervals.length; i++) {
      if (intervals[i][0] - previousEndTime > 0) {
        output.add(new int[] { previousStartTime, previousEndTime });
        previousStartTime = intervals[i][0];
        previousEndTime = intervals[i][1];
      } else {
        previousEndTime = Math.max(intervals[i][1], previousEndTime);
      }
    }
    output.add(new int[] { previousStartTime, previousEndTime });
    return output.toArray(new int[output.size()][]);
  }
}